processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -441,6 +441,19 @@ public InputRow formatRow(InputRow row)
     return builder.build();
   }
 
+  @Nullable
+  @Override
+  public ColumnCapabilities getColumnCapabilities(String columnName)
+  {
+    if (timeAndMetricsColumnCapabilities.containsKey(columnName)) {
+      return timeAndMetricsColumnCapabilities.get(columnName);
+    }
+    if (dimensionDescs.containsKey(columnName)) {
```

Comment:
Looking through the code in IncrementalIndex, sometimes `dimensionDescs` is accessed with synchronization.  Sometimes it is not.  Assuming that it's okay to access it without synchronization, this looks good.  Have we verified that we will already be in a synchronized block once we get to this method?

Reply:
Good question, will have a look. `timeAndMetricsColumnCapabilities` would also have this problem potentially

Reply:
hmm, I think it should be synchronized anyway because callers would not expect this method to not be threadsafe. it doesn't seem to cause a performance impact afaict:

without:
![Screen Shot 2021-12-09 at 3 30 16 PM](https://user-images.githubusercontent.com/1577461/145491925-fb52f4b9-93b7-4616-bf9a-0b1a4b3630c3.png)


with:
![Screen Shot 2021-12-09 at 3 30 04 PM](https://user-images.githubusercontent.com/1577461/145491931-39672b36-e262-471b-beb8-490ccef69477.png)

Reply:
btw, I didn't trust any of the users of `dimensionDescs` that weren't obviously safe so adjusted a couple of other locations as well before I did the measurement in my previous comment